### PR TITLE
ResetBlockFailureFlags did not remove the invalidity flag in other chain

### PR DESCRIPTION
**Pull request by ghost**
I think that I may have found a bug in ResetBlockFailureFlags.
The bug happens when there are two chains (say two blocks with blockheight 200) and a block before that fork will be invalidated (say blockheight 100).
If you then use ResetBlockFailureFlags to remove the failure flags from one of the two blocks at block height 200.
ResetBlockFailureFlags will remove the flags of all descendants and ancestors of the block 200 on one of the chains.
But for the block number 200 on the other chain and the following blocks the flags will not be removed.
The BLOCK_FAILED_CHILD flags remain for these blocks. But the BLOCK_FAILED_VALID flag of block 100 will be removed.
This can then produce a problem in CheckBlockIndex on line

assert((pindex->nStatus & BLOCK_FAILED_MASK) == 0); // The failed mask cannot be set for blocks without invalid parents.

My fix searches for a BLOCK_FAILED_VALID flag and remembers the block index. It will then call ResetBlockFailureFlags for this block index again to remove the flags all all descendants.

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -3093,16 +3093,26 @@ void CChainState::ResetBlockFailureFlags(CBlockIndex *pindex) {
...
+				pindexFirstInvalid = pindex;
+			}
             pindex->nStatus &= ~BLOCK_FAILED_MASK;
             setDirtyBlockIndex.insert(pindex);
             m_blockman.m_failed_blocks.erase(pindex);
         }
         pindex = pindex->pprev;
     }
+	
+	//Remove the invalidity flag from all descendants of the first invalid ancestor block
```
Comment (steevo87) [inline src/validation.cpp:3112]: Changing tab to spaces:

```suggestion
   //Remove the invalidity flag from all descendants of the first invalid ancestor block
```

File: src/validation.cpp [hunk 2]
```diff
@@ -3093,16 +3093,26 @@ void CChainState::ResetBlockFailureFlags(CBlockIndex *pindex) {
         }
         it++;
     }
+	
+	CBlockIndex *pindexFirstInvalid = nullptr;
```
Comment (steevo87) [inline src/validation.cpp:3097]: Changing tab to spaces:

```suggestion
   CBlockIndex *pindexFirstInvalid = nullptr;
```

File: src/validation.cpp [hunk 3]
```diff
@@ -3093,16 +3093,26 @@ void CChainState::ResetBlockFailureFlags(CBlockIndex *pindex) {
...
 
     // Remove the invalidity flag from all ancestors too.
     while (pindex != nullptr) {
         if (pindex->nStatus & BLOCK_FAILED_MASK) {
+			if (pindex->nStatus & BLOCK_FAILED_VALID) {
+				pindexFirstInvalid = pindex;
+			}
```
Comment (steevo87) [inline src/validation.cpp:3104]: Changing tabs to spaces:

```suggestion
         if (pindex->nStatus & BLOCK_FAILED_VALID) {
            pindexFirstInvalid = pindex;
         }
```

**Comment by MarcoFalke** [comment]
I haven't looked at the diff, but it would be nice to have a test that fails before the changes and passes after.

**Comment by fanquake** [comment]
Closing for now (no follow up, dead GH account), and marking up for grabs. As mentioned, a test showing the bug would also be a good place to start.

**Pull request closed without merging**